Create an example that includes tests and how to run them. #122

hmelenok opened the pull request:
Please check #122

wbamberg commented:
@hmelenok , nice work!

@andymckay , do you know who might be qualified to review this? There's a lot of stuff in here I don't know, so while I could probably figure out whether it's working properly, I probably won't be able to tell whether it's good practice.

andymckay commented:
@kumar303 might be able to help.

andymckay commented on mocha-client-tests/addon/manifest.json:
```diff
@@ -0,0 +1,53 @@
+  },
+  "applications": {
+    "gecko": {
+      "id": "production@gemshelf.com",
```
We've been trying to remove the id out of example extensions, they can be through temporary install locally or pushed to AMO for signing. If there's no id, they won't all conflict with each other.

hmelenok replied:
Copy paste mistake

hmelenok replied:
Please check it

kumar303 commented on mocha-client-tests/tests/lib/background.test.js:
```diff
@@ -0,0 +1,9 @@
+    describe('ping', function() {
+        it('should return pong in response', function() {
+            Background.ping(false, false, function(response) {
+                response.should.equal('pong');
```
chai's prototype hack might be confusing to the casual JavaScript reader. Could you update this to use [expect()](http://chaijs.com/api/bdd/) (or something) instead? It would look like `expect(response).to.equal('pong')`

hmelenok replied:
Thanks, changed to expect 👍

kumar303 replied:
I don't see the change to make this `expect()`. Do you have any unpushed commits maybe?

kumar303 commented on mocha-client-tests/addon/tests/lib/background-messaging.test.js:
```diff
@@ -0,0 +1,13 @@
+ */
+describe('Background', function() {
+    describe('ping', function() {
+        it('should return pong in response', function(done) {
```
how do I run this test? It isn't executed with `npm run test` from the root directory. Is there a different way to run it?

hmelenok replied:
`npm run test` - runs phantomjs tests, there UI ob mocha which included as example if you install all dependencies, please read README.md before

kumar303 commented on mocha-client-tests/README.md:
```diff
@@ -0,0 +1,13 @@
+##PhantomJs tests
+`npm run test` will run simple test  `./addon/background.js`
+## Mocha browser test
+Already inside of addon 
```
What do you mean by this? I couldn't figure out how to run the tests in `addon/tests/lib`. How do I run them and see results?

hmelenok replied:
`cd mocha-client-tests/addon && npm i` - it will install all node modules which required for running tests in addon window

hmelenok replied:
Then just run `web-ext run  -s ./addon` with cli

kumar303 replied:
This "Mocha browser test" section is redundant, right? If so, I suggest removing it. It seems to be identical to "Run with web-ext cli."

hmelenok commented on mocha-client-tests/addon/package.json:
```diff
@@ -0,0 +1,16 @@
+    "mocha": "^3.1.2",
+    "should": "^11.1.1"
+  }
+}
```
There two type of tests inside of addon i had add mocha client tests

hmelenok replied:
Outside of addon folder there phantomjs tests

hmelenok commented on mocha-client-tests/addon/scripts/popup.js:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Created by hmelenok on 10/30/16.
```
Done removed

hmelenok commented:
Please check last changes: 

- removed comments from js files
- removed jquery and empty contentscript 

<img width="444" alt="screen shot 2016-11-05 at 19 42 53" src="https://cloud.githubusercontent.com/assets/2149114/20032307/6ded3a9e-a390-11e6-8d87-216853dba0f5.png">

kumar303 commented on mocha-client-tests/addon/tests/lib/test.array.js:
```diff
@@ -0,0 +1,8 @@
+describe('Array', function() {
+    describe('#indexOf()', function() {
+        it('should return -1 when the value is not present', function() {
+            [1,2,3].indexOf(5).should.equal(-1);
```
This fails with `should is undefined`; it needs to use `expect()` instead, right?

![screenshot 2016-11-07 10 59 25](https://cloud.githubusercontent.com/assets/55398/20067094/72b9baee-a4d9-11e6-8d32-36b2e7104306.png)

kumar303 commented on mocha-client-tests/README.md:
```diff
@@ -0,0 +1,13 @@
+And inside of addon please run `npm i`
+
+##Run with web-ext cli
+If you have it already run `npm run web-ext` if not `npm i web-ext -g` and then first command (will work with FF dev edition), if you have error with web-ext cli please add path for FF binary file with `--firefox-binary /path/to/firefox-bin`
```
`npm run web-ext` doesn't work because `package.json` wants it to be `npm run web-ext-run`. I think you should rename the `script` in `package.json` to `web-ext`.

kumar303 replied:
`npm install -g web-ext` is not necessary because you've already put it in the `devDependencies`. When you do `npm run` it figures out the path to `web-ext` automatically.

kumar303 replied:
These instructions need a final sentence explaining that you need to click the mocha icon from the browser bar to run the tests

kumar303 commented on mocha-client-tests/README.md:
```diff
@@ -0,0 +1,13 @@
+##Install dependency
+`npm install` and `bower install`  
+
+And inside of addon please run `npm i`
```
I think it would be more helpful to spell out `npm install` in all cases since not all readers will know about the `i` alias.

hmelenok replied:
Done

kumar303 commented on mocha-client-tests/README.md:
```diff
@@ -0,0 +1,13 @@
+If you have it already run `npm run web-ext` if not `npm i web-ext -g` and then first command (will work with FF dev edition), if you have error with web-ext cli please add path for FF binary file with `--firefox-binary /path/to/firefox-bin`
+[web-ext docs](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/web-ext_command_reference)
+##PhantomJs tests
+`npm run test` will run simple test  `./addon/background.js`
```
This can just be `npm test` which is a more standard way to run it using `npm`.

hmelenok replied:
Done

kumar303 commented on mocha-client-tests/addon/tests/lib/background-messaging.test.js:
```diff
@@ -0,0 +1,10 @@
+describe('Background', function() {
+    describe('ping', function() {
+        it('should return pong in response', function(done) {
+            chrome.runtime.sendMessage({action: 'ping'},function(response) {
```
to keep the style consistent, add a space before `function`

hmelenok replied:
I'm not sure by which style (as i don't have spaces anywhere)

kumar303 commented on mocha-client-tests/addon/background.js:
```diff
@@ -0,0 +1,13 @@
+    receiveMessage: function(msg, sender, sendResponse) {
+        if (msg && msg.action && Background.hasOwnProperty(msg.action)) {
+            return Background[msg.action](msg, sender, sendResponse);
+        }
```
it would be nice if this had an else clause to warn about missed messages. It could help someone who may be editing this code. Something like:
````
... else {
    console.warning('No handler for message: ' + JSON.stringify(msg));
}
````

kumar303 commented on mocha-client-tests/addon/tests/lib/background-messaging.test.js:
```diff
@@ -0,0 +1,10 @@
+    describe('ping', function() {
+        it('should return pong in response', function(done) {
+            chrome.runtime.sendMessage({action: 'ping'},function(response) {
+                expect(response).to.equal('pong');
```
when this assertion fails, it only gets logged to the browser console. Can you make it show up in the mocha test output instead? For example, if you change `expect(response).to.equal('pong')` to `expect(response).to.equal('nope')` then I see this confusing timeout instead of the assertion error:

![screenshot 2016-11-07 11 25 33](https://cloud.githubusercontent.com/assets/55398/20068082/1514c772-a4dd-11e6-99b3-5130cb4afb9a.png)

hmelenok replied:
fixed

hmelenok replied:
https://github.com/mochajs/mocha/issues/1128 - looks like I'm not sure why it wait until fail

kumar303 replied:
ok, I see what the issue is now. The `expect()` call throws an error but that is not passed to `done(error)` so the mocha reporter can't show it. The way to fix this is to rewrite the code using promises:
````javascript
it('should return pong in response', function() {
  // Return a promise for Mocha using the Firefox browser API instead of chrome.
  return browser.runtime.sendMessage({action: 'ping'})
    .then(function(response) {
      expect(response).to.equal('pong');
    });
});
````
I think it would be good to convert this code into promises since the developer can always use a [polyfill](https://github.com/mozilla/webextension-polyfill) to make it compatible with Chrome.

hmelenok replied:
Thanks for advice it works!

hmelenok replied:
Added in last commit

kumar303 commented on mocha-client-tests/addon/manifest.json:
```diff
@@ -0,0 +1,38 @@
+    "default_title": "Mocha Test",
+    "default_popup": "popup.html"
+  },
+  "applications": {
```
You can remove this section entirely. It's not needed for this example. It could be harmful if someone copy/pasted it because it makes the extension only compatible with Firefox, not Chrome.

hmelenok replied:
done

kumar303 commented on mocha-client-tests/addon/manifest.json:
```diff
@@ -0,0 +1,38 @@
+      "background.js"
+    ]
+  },
+  "permissions": [
```
This looks like it has a bunch of permissions not needed for this example. Can you strip it down to only the relevant permissions? I think only `notifications` is needed, right?

kumar303 commented on mocha-client-tests/addon/tests/mocha-run.js:
```diff
@@ -0,0 +1,3 @@
+mocha.checkLeaks();
+mocha.globals(['jQuery','AppView', 'ExtensionOptions', 'ExtensionView', 'WebView']);
```
where are these globals used? These don't look like they are relevant to the example

hmelenok replied:
There a list of default globals in extension which from Mocha perspective looks like global leaks

kumar303 commented on mocha-client-tests/package.json:
```diff
@@ -0,0 +1,21 @@
+  "license": "MPL-2.0",
+  "devDependencies": {
+    "chai": "^3.5.0",
+    "chrome-mock": "0.0.9",
```
where is this used?

hmelenok replied:
chrome-mock used for phantomJS as it not have WebExtensions (or Chrome Extensions) API inside of engine, so i'm using mock of this API's. As when you include to test runner original file it throws errors about chrome is not defined

kumar303 replied:
aha, I see

kumar303 commented:
I see that you pushed some more changes (thanks!). Just let me know when it's ready for another review.

hmelenok commented:
Please check it again)

kumar303 commented on mocha-client-tests/README.md:
```diff
@@ -0,0 +1,17 @@
+
+When addon will start click on mocha icon in your browser bar to run client tests: 
+
+![addon screenshot](screenshots/addon-button.png "Mocha test addon")
```
good idea to add a screenshot

hmelenok replied:
👍

kumar303 commented on mocha-client-tests/addon/scripts/browser-polyfill.js:
```diff
@@ -0,0 +1,849 @@
+/* webextension-polyfill - v0.1.0 - Wed Nov 09 2016 16:37:59 */
```
this is the built dist file, right? I don't know what to look for.

hmelenok replied:
That what i get after build in `webextension-polyfill`

kumar303 commented:
Thanks for all the fixups!
